server/client.go

```diff
@@ -3325,12 +3334,17 @@ func (c *client) pubAllowedFullCheck(subject string, fullCheck bool) bool {
 				allowed = true
 			}
 		}
-		c.mu.Unlock()
+		if !hasLock {
+			c.mu.Unlock()
+		}
 	} else {
 		// Update our cache here.
-		c.perms.pcache[string(subject)] = allowed
-		// Prune if needed.
-		if len(c.perms.pcache) > maxPermCacheSize {
+		c.perms.pcache.Store(string(subject), allowed)
+		// There is a case where we can invoke this from multiple go routines,
+		// (in deliverMsg() if sub.client is a LEAF), so we make sure to prune
+		// from only one go routine at a time.
+		if n := atomic.AddInt32(&c.perms.pcsz, 1); n > maxPermCacheSize &&
+			atomic.CompareAndSwapInt32(&c.perms.prun, 0, 1) {
```

Comment:
I wonder if the compare and swap portion here should move into prunePubPermsCache, where we also clear the value.
```
if !atomic.CompareAndSwapInt32(&c.perms.prun, 0, 1) {
    return
}
```

Reply:
Agreed.